From 27fdc7497ff900d1d59d8f3e5c0415e0f89e5eab Mon Sep 17 00:00:00 2001
From: Sjoerd Simons <sjoerd.simons@collabora.co.uk>
Date: Thu, 13 Nov 2014 13:09:34 +0100
Subject: [PATCH 120/121] script: Add SetRefreshRate function

The script plugin hardcodes an FPS value of 50, which for some themes
and various devices is a bit much.  Add a new function
(Plymouth.SetRefreshRate) which sets the rate at which the
RefreshFunction gets called so each script theme can determine their most
appropriate rate.

https://bugs.freedesktop.org/show_bug.cgi?id=86247
---
 src/plugins/splash/script/plugin.c              |  5 +++--
 src/plugins/splash/script/script-lib-plymouth.c | 19 ++++++++++++++++++-
 src/plugins/splash/script/script-lib-plymouth.h |  4 +++-
 3 files changed, 24 insertions(+), 4 deletions(-)

diff --git a/src/plugins/splash/script/plugin.c b/src/plugins/splash/script/plugin.c
index 1612efd..6484766 100644
--- a/src/plugins/splash/script/plugin.c
+++ b/src/plugins/splash/script/plugin.c
@@ -220,7 +220,7 @@ on_timeout (ply_boot_splash_plugin_t *plugin)
 {
         double sleep_time;
 
-        sleep_time = 1.0 / FRAMES_PER_SECOND;
+        sleep_time = 1.0 / plugin->script_plymouth_lib->refresh_rate;
         ply_event_loop_watch_for_timeout (plugin->loop,
                                           sleep_time,
                                           (ply_event_loop_timeout_handler_t)
@@ -272,7 +272,8 @@ start_script_animation (ply_boot_splash_plugin_t *plugin)
         plugin->script_sprite_lib = script_lib_sprite_setup (plugin->script_state,
                                                              plugin->displays);
         plugin->script_plymouth_lib = script_lib_plymouth_setup (plugin->script_state,
-                                                                 plugin->mode);
+                                                                 plugin->mode,
+                                                                 FRAMES_PER_SECOND);
         plugin->script_math_lib = script_lib_math_setup (plugin->script_state);
         plugin->script_string_lib = script_lib_string_setup (plugin->script_state);
 
diff --git a/src/plugins/splash/script/script-lib-plymouth.c b/src/plugins/splash/script/script-lib-plymouth.c
index 16806cd..857792b 100644
--- a/src/plugins/splash/script/script-lib-plymouth.c
+++ b/src/plugins/splash/script/script-lib-plymouth.c
@@ -48,6 +48,15 @@ static script_return_t plymouth_set_function (script_state_t *state,
         return script_return_obj_null ();
 }
 
+static script_return_t plymouth_set_refresh_rate (script_state_t *state,
+                                                  void           *user_data)
+{
+      script_lib_plymouth_data_t *data = user_data;
+      data->refresh_rate = script_obj_hash_get_number (state->local, "value");
+
+      return script_return_obj_null ();
+}
+
 static script_return_t plymouth_get_mode (script_state_t *state,
                                           void           *user_data)
 {
@@ -73,7 +82,8 @@ static script_return_t plymouth_get_mode (script_state_t *state,
 }
 
 script_lib_plymouth_data_t *script_lib_plymouth_setup (script_state_t        *state,
-                                                       ply_boot_splash_mode_t mode)
+                                                       ply_boot_splash_mode_t mode,
+                                                       int refresh_rate)
 {
         script_lib_plymouth_data_t *data = malloc (sizeof(script_lib_plymouth_data_t));
 
@@ -90,6 +100,7 @@ script_lib_plymouth_data_t *script_lib_plymouth_setup (script_state_t        *st
         data->script_quit_func = script_obj_new_null ();
         data->script_system_update_func = script_obj_new_null ();
         data->mode = mode;
+        data->refresh_rate = refresh_rate;
 
         script_obj_t *plymouth_hash = script_obj_hash_get_element (state->global, "Plymouth");
         script_add_native_function (plymouth_hash,
@@ -99,6 +110,12 @@ script_lib_plymouth_data_t *script_lib_plymouth_setup (script_state_t        *st
                                     "function",
                                     NULL);
         script_add_native_function (plymouth_hash,
+                                    "SetRefreshRate",
+                                    plymouth_set_refresh_rate,
+                                    data,
+                                    "value",
+                                    NULL);
+        script_add_native_function (plymouth_hash,
                                     "SetBootProgressFunction",
                                     plymouth_set_function,
                                     &data->script_boot_progress_func,
diff --git a/src/plugins/splash/script/script-lib-plymouth.h b/src/plugins/splash/script/script-lib-plymouth.h
index bf8c96e..0d3fe66 100644
--- a/src/plugins/splash/script/script-lib-plymouth.h
+++ b/src/plugins/splash/script/script-lib-plymouth.h
@@ -41,10 +41,12 @@ typedef struct
         script_obj_t          *script_quit_func;
         script_obj_t           *script_system_update_func;
         ply_boot_splash_mode_t mode;
+        int                    refresh_rate;
 } script_lib_plymouth_data_t;
 
 script_lib_plymouth_data_t *script_lib_plymouth_setup (script_state_t        *state,
-                                                       ply_boot_splash_mode_t mode);
+                                                       ply_boot_splash_mode_t mode,
+                                                       int refresh_rate);
 void script_lib_plymouth_destroy (script_lib_plymouth_data_t *data);
 
 void script_lib_plymouth_on_refresh (script_state_t             *state,
-- 
1.9.0

